README.md
# Calendar.strftime options generator for CLDR

[Calendar.strftime/3](https://hexdocs.pm/elixir/Calendar.html#strftime/3) has been available since Elixir 1.11 to provide date/datetime formatting using the principles that date back to at least [1978](https://en.wikipedia.org/wiki/The_C_Programming_Language).

The functions in this library are intended to serve options to `Calendar.strftime/3` to support localisation of dates/datetimes leveraging the content in [CLDR](https://cldr.unicode.org). Therefore a developer can take advantage of the built-in localised formats from `CLDR` with well-known `strftime` formatting strings.

### Installation

The package can be installed by adding `cldr_strftime` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:ex_cldr_strftime, "~> 0.2.0"}
  ]
end
```

### Configuration

Update your `ex_cldr` backend module to include the provider module `Cldr.Strftime`. For example:

```elixir
defmodule MyApp.Cldr do
  use Cldr,
    locales: ["en", "fr", "af", "ja", "de", "pl", "th"],
    providers: [Cldr.Number, Cldr.Calendar, Cldr.DateTime, Cldr.Strftime]
end
```

Documentation is available at [https://hexdocs.pm/cldr_strftime](https://hexdocs.pm/cldr_strftime).

### Keyword Options returned

In accordance with the options defined for `Calendar.strftime/3`, `Cldr.Strftime.strftime_options!/2` returns the following keyword list:

* `:preferred_datetime` - a string for the preferred format to show datetimes,

* `:preferred_date` - a string for the preferred format to show dates.

* `:preferred_time` - a string for the preferred format to show times.

* `:am_pm_names` - a function that receives either :am or :pm and returns the name of the period of the day

* `:month_names` - a function that receives a number and returns the name of the corresponding month.

* `:abbreviated_month_names` - a function that receives a number and returns the abbreviated name of the corresponding month.

* `:day_of_week_names` - a function that receives a number and returns the name of the corresponding day of week.

* `:abbreviated_day_of_week_names` - a function that receives a number and returns the abbreviated name of the corresponding day of week

### CLDR format translation

[CLDR](https://cldr.unicode.org) [formats dates, times and date times](https://unicode.org/reports/tr35/tr35-dates.html#Contents) using a [different formatting system](https://unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table) to that of `Calendar.strftime/3`.  The CLDR formats are translated at compile time according to the following table. Since them formats are translated at compile time, performance is comparable to using native `Calendar.strftime/3`formats natively.

Strftime | CLDR     | Description                                                   | Examples (in ISO)
-------- | ---------| ------------------------------------------------------------- | ------------------
a        | E,EE,EEE | Abbreviated name of day                                       | Mon
A        | EEEE     | Full name of day                                              | Monday
b        | MMM      | Abbreviated month name                                        | Jan
B        | MMMM     | Full month name                                               | January
d        | d        | Day of the month                                              | 01, 12
H        | h        | Hour using a 24-hour clock                                    | 00, 23
I        | H        | Hour using a 12-hour clock                                    | 01, 12
j        | DDD      | Day of the year                                               | 001, 366
m        | MM       | Month                                                         | 01, 12
M        | mm       | Minute                                                        | 00, 59
p        | a,aa,aaa | "AM" or "PM" (noon is "PM", midnight as "AM")                 | AM, PM
q        | Q        | Quarter                                                       | 1, 2, 3, 4
S        | ss       | Second                                                        | 00, 59, 60
u        | s        | Day of the week                                               | 1 (Monday), 7 (Sunday)
y        | YY       | Year as 2-digits                                              | 01, 01, 86, 18
Y        | YYYY     | Year                                                          | -0001, 0001, 1986
z        | ZZZZ     | +hhmm/-hhmm time zone offset from UTC (empty string if naive) | +0300, -0530
Z        | V, VV    | Time zone abbreviation (empty string if naive)                | CET, BRST

[Calendar.strftime/3](https://hexdocs.pm/elixir/Calendar.html#strftime/3) allows for a set of options to guide formatting

### Example usage

These examples use the `%c`, `%x` and `%X` format flags which means "use the preferred_* option if provided".

```elixir
iex> Calendar.strftime ~U[2019-08-26 13:52:06.0Z], "%x", Cldr.Strftime.strftime_options!()
"Aug 26, 2019"

iex> Calendar.strftime ~U[2019-08-26 13:52:06.0Z], "%X", Cldr.Strftime.strftime_options!()
"13:52:06 PM"

iex> Calendar.strftime ~U[2019-08-26 13:52:06.0Z], "%c", Cldr.Strftime.strftime_options!()
"Aug 26, 2019, 13:52:06 PM"

iex> Calendar.strftime ~U[2019-08-26 13:52:06.0Z], "%c", Cldr.Strftime.strftime_options!("ja")
"2019/08/26 13:52:06"

iex> Calendar.strftime ~U[2019-08-26 13:52:06.0Z], "%c", Cldr.Strftime.strftime_options!("ja", format: :long)
"2019年08月26日 13:52:06 +0000"

iex> Calendar.strftime ~U[2019-08-26 13:52:06.0Z], "%c", Cldr.Strftime.strftime_options!("pl")
"26 sie 2019, 13:52:06"

iex> Calendar.strftime ~U[2019-08-26 13:52:06.0Z], "%c", Cldr.Strftime.strftime_options!("pl", format: :long)
"26 sierpnia 2019 13:52:06 +0000"
```




CHANGELOG.md
# Changelog

## Cldr_Strftime v1.0.1

This is the changelog for Cldr_Strftime v1.0.1 released on September 30th, 2023.  For older changelogs please consult the release tag on [GitHub](https://github.com/elixir-cldr/cldr_strftime/tags)

### Bug Fixes

* Fixes code ambiguity by wrapping parens. Thanks to @rubas for the report. Closes #2.

## Cldr_Strftime v1.0.0

This is the changelog for Cldr_Strftime v1.0.0 released on December 1st, 2021.  For older changelogs please consult the release tag on [GitHub](https://github.com/elixir-cldr/cldr_strftime/tags)

### Enhancements

* Bump to version 1.0.0 since the API has been stable for over a year

* Update requirement for CLDR40-based releases of `ex_cldr` to remove deprecation warnings.  Thanks to @zenbaku for the PR.

## Cldr_Strftime v0.2.0

This is the changelog for Cldr_Strftime v0.2.0 released on March 18th, 2021.  For older changelogs please consult the release tag on [GitHub](https://github.com/elixir-cldr/cldr_strftime/tags)

### Enhancements

* Update documentation and benchmarks

## Cldr_Strftime v0.1.0

This is the changelog for Cldr_Strftime v0.1.0 released on November 29th, 2020.  For older changelogs please consult the release tag on [GitHub](https://github.com/elixir-cldr/cldr_strftime/tags)

### Enhancements

* Initial release

test/cldr_strftime_test.exs
defmodule Cldr.Strftime.Test do
  use ExUnit.Case
  doctest Cldr.Strftime

end


mix.exs
defmodule Cldr.Strftime.MixProject do
  use Mix.Project

  @version "1.0.1"

  def project do
    [
      app: :ex_cldr_strftime,
      version: @version,
      name: "Cldr Strftime Options generator for Calendar.strftime/3",
      source_url: "https://github.com/elixir-cldr/cldr_strftime",
      docs: docs(),
      elixir: "~> 1.10",
      description: description(),
      package: package(),
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      compilers: Mix.compilers(),
      elixirc_paths: elixirc_paths(Mix.env()),
      dialyzer: [
        ignore_warnings: ".dialyzer_ignore_warnings",
        plt_add_apps: ~w(calendar_interval)a
      ]
    ]
  end

  defp description do
    """
    Generates a Calendar.strftime/3 options list
    from CLDR date time formats to support localised date, time
    and date time formatting.
    """
  end

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  def docs do
    [
      source_ref: "v#{@version}",
      main: "readme",
      extras: ["README.md", "CHANGELOG.md", "LICENSE.md"],
      logo: "logo.png",
      skip_undefined_reference_warnings_on: ["changelog", "readme", "CHANGELOG.md"]
    ]
  end

  defp deps do
    [
      {:ex_cldr_numbers, "~> 2.28"},
      {:ex_cldr_calendars, "~> 1.17"},
      {:ex_cldr_dates_times, "~> 2.10"},
      {:calendar_interval, "~> 0.2", optional: true},
      {:ex_doc, "~> 0.18", optional: true, runtime: false},
      {:jason, "~> 1.0", optional: true},
      {:benchee, "~> 1.0", optional: true, only: :dev, runtime: false},
      {:dialyxir, "~> 1.0", only: [:dev], runtime: false}
    ]
  end

  defp package do
    [
      maintainers: ["Kip Cole"],
      licenses: ["Apache-2.0"],
      links: links(),
      files: [
        "lib",
        "config",
        "mix.exs",
        "README*",
        "CHANGELOG*",
        "LICENSE*"
      ]
    ]
  end

  def links do
    %{
      "GitHub" => "https://github.com/elixir-cldr/cldr_strftime",
      "Changelog" =>
        "https://github.com/elixir-cldr/cldr_strftime/blob/v#{@version}/CHANGELOG.md",
      "Readme" => "https://github.com/elixir-cldr/cldr_strftime/blob/v#{@version}/README.md"
    }
  end

  defp elixirc_paths(:test), do: ["lib", "mix", "test"]
  defp elixirc_paths(:dev), do: ["lib", "mix"]
  defp elixirc_paths(:docs), do: ["lib", "mix"]
  defp elixirc_paths(_), do: ["lib"]
end


lib/cldr/strftime/backend.ex
defmodule Cldr.Strftime.Backend do
  @moduledoc false

  def define_backend_modules(config) do
    backend = config.backend
    config = Macro.escape(config)

    quote location: :keep, bind_quoted: [config: config, backend: backend] do
      defmodule Strftime do
        @moduledoc false

        alias Cldr.Calendar
        alias Cldr.LanguageTag

        def date_formats(locale \\ Cldr.get_locale())

        def date_formats(%LanguageTag{cldr_locale_name: cldr_locale_name}) do
          date_formats(cldr_locale_name)
        end

        def time_formats(locale \\ Cldr.get_locale())

        def time_formats(%LanguageTag{cldr_locale_name: cldr_locale_name}) do
          time_formats(cldr_locale_name)
        end

        def date_time_formats(locale \\ Cldr.get_locale())

        def date_time_formats(%LanguageTag{cldr_locale_name: cldr_locale_name}) do
          date_time_formats(cldr_locale_name)
        end

        for locale <- Cldr.Locale.Loader.known_locale_names(config) do
          locale_data = Cldr.Locale.Loader.get_locale(locale, config)
          calendar_data = get_in(locale_data, [:dates, :calendars, :gregorian])

          date_formats =
            calendar_data
            |> Map.fetch!(:date_formats)
            |> Cldr.Strftime.Translator.translate()

          def date_formats(unquote(locale)) do
            {:ok, unquote(Macro.escape(date_formats))}
          end

          time_formats =
            calendar_data
            |> Map.fetch!(:time_formats)
            |> Cldr.Strftime.Translator.translate()

          def time_formats(unquote(locale)) do
            {:ok, unquote(Macro.escape(time_formats))}
          end

          date_time_formats =
            calendar_data
            |> Map.fetch!(:date_time_formats)
            |> Cldr.Strftime.Translator.substitute(date_formats, time_formats)

          def date_time_formats(unquote(locale)) do
            {:ok, unquote(Macro.escape(date_time_formats))}
          end
        end
      end

      def date_formats(locale), do: {:error, Cldr.Locale.locale_error(locale)}
      def time_formats(locale), do: {:error, Cldr.Locale.locale_error(locale)}
      def date_time_formats(locale), do: {:error, Cldr.Locale.locale_error(locale)}
    end
  end
end


lib/cldr/strftime/translator.ex
defmodule Cldr.Strftime.Translator do
  @moduledoc false

  @standard_formats [:short, :medium, :long, :full]

  def substitute(date_time_formats, date_formats, time_formats) do
    Enum.reduce @standard_formats, Map.new(), fn format, acc ->
      datetime_format = Map.fetch!(date_time_formats, format)
      date_format = Map.fetch!(date_formats, format)
      time_format = Map.fetch!(time_formats, format)

      parsed =
        datetime_format
        |> Cldr.Substitution.parse
        |> Enum.map(&translate/1)

      new_datetime_format =
        [time_format, date_format]
        |> Cldr.Substitution.substitute(parsed)
        |> Enum.join

      Map.put(acc, format, new_datetime_format)
    end
  end

  def translate(formats) when is_map(formats) do
    formats
    |> Map.take(@standard_formats)
    |> Enum.map(fn {k, v} -> {k, translate(v)} end)
    |> Map.new
  end

  def translate("") do
    ""
  end

  def translate(<< "\\%", rest :: binary>>) do
    << "%%", translate(rest) :: binary >>
  end

  def translate(<< "\\", char :: utf8, rest :: binary>>) do
    << "\\", char :: utf8, translate(rest) :: binary >>
  end

  def translate("'" <> rest) do
    quoted_string(rest)
  end

  def translate("%" <> rest) do
    "%%" <> translate(rest)
  end

  def translate("EEEE" <> rest) do
    "%A" <> translate(rest)
  end

  def translate("EEE" <> rest) do
    "%a" <> translate(rest)
  end

  def translate("EE" <> rest) do
    "%a" <> translate(rest)
  end

  def translate("E" <> rest) do
    "%a" <> translate(rest)
  end

  def translate("MMMM" <> rest) do
    "%B" <> translate(rest)
  end

  def translate("MMM" <> rest) do
    "%b" <> translate(rest)
  end

  def translate("dddd" <> rest) do
    "%4d" <> translate(rest)
  end

  def translate("ddd" <> rest) do
    "%3d" <> translate(rest)
  end

  def translate("dd" <> rest) do
    "%2d" <> translate(rest)
  end

  def translate("d" <> rest) do
    "%d" <> translate(rest)
  end

  def translate("hh" <> rest) do
    "%2H" <> translate(rest)
  end

  def translate("h" <> rest) do
    "%H" <> translate(rest)
  end

  def translate("HH" <> rest) do
    "%2H" <> translate(rest)
  end

  def translate("H" <> rest) do
    "%H" <> translate(rest)
  end

  def translate("DDD" <> rest) do
    "%j" <> translate(rest)
  end

  def translate("MM" <> rest) do
    "%m" <> translate(rest)
  end

  def translate("M" <> rest) do
    "%m" <> translate(rest)
  end

  def translate("mm" <> rest) do
    "%M" <> translate(rest)
  end

  def translate("aaa" <> rest) do
    "%p" <> translate(rest)
  end

  def translate("aa" <> rest) do
    "%p" <> translate(rest)
  end

  def translate("a" <> rest) do
    "%p" <> translate(rest)
  end

  def translate("Q" <> rest) do
    "%q" <> translate(rest)
  end

  def translate("ss" <> rest) do
    "%S" <> translate(rest)
  end

  def translate("s" <> rest) do
    "%u" <> translate(rest)
  end

  def translate("YYYY" <> rest) do
    "%Y" <> translate(rest)
  end

  def translate("YY" <> rest) do
    "%y" <> translate(rest)
  end

  def translate("yyyy" <> rest) do
    "%4Y" <> translate(rest)
  end

  def translate("yy" <> rest) do
    "%2Y" <> translate(rest)
  end

  def translate("y" <> rest) do
    "%Y" <> translate(rest)
  end

  def translate("ZZZZ" <> rest) do
    "%Z" <> translate(rest)
  end

  def translate("ZZZ" <> rest) do
    "%Z" <> translate(rest)
  end

  def translate("ZZ" <> rest) do
    "%Z" <> translate(rest)
  end

  def translate("Z" <> rest) do
    "%Z" <> translate(rest)
  end

  def translate("zzzz" <> rest) do
    "%z" <> translate(rest)
  end

  def translate("zzz" <> rest) do
    "%z" <> translate(rest)
  end

  def translate("zz" <> rest) do
    "%z" <> translate(rest)
  end

  def translate("z" <> rest) do
    "%z" <> translate(rest)
  end

  def translate("VV" <> rest) do
    "%Z" <> translate(rest)
  end

  def translate("V" <> rest) do
    "%Z" <> translate(rest)
  end

  def translate(<< char :: utf8, rest :: binary >>) do
    << char :: utf8, translate(rest) :: binary >>
  end

  def translate(other) do
    other
  end

  defp quoted_string(string, acc \\ "")

  defp quoted_string("", acc) do
    acc
  end

  defp quoted_string("'" <> rest, acc) do
    acc <> rest
  end

  defp quoted_string(<< "\\", char :: utf8, rest :: binary >>, acc) do
    quoted_string(rest, << acc :: binary, char :: utf8 >>)
  end

  defp quoted_string(<< char :: utf8, rest :: binary >>, acc) do
    quoted_string(rest, << acc :: binary, char :: utf8 >>)
  end

end

lib/cldr_strftime.ex
defmodule Cldr.Strftime do

  @doc false
  def cldr_backend_provider(config) do
    Cldr.Strftime.Backend.define_backend_modules(config)
  end

  @doc """
  Returns a keyword list of options than can be applied to
  `NimbleStrftime.format/3`.

  The hex package [nimble_strftime](https://hex.pm/packages/nimble_strftime)
  provides a `format/3` function to format dates, times and datetimes.
  It takes a set of options that can return day, month and am/pm names.

  `strftime_options!` returns a keyword list than can be used as these
  options to return localised names for days, months and am/pm.

  ## Arguments

  * `locale` is any locale returned by `Cldr.known_locale_names/1`. The
    default is `Cldr.get_locale/0`

  * `options` is a set of keyword options. The default is `[]`

  ## Options

  * `:format` determines the format of the date, time or
    date time. The options are `:short`, `:medium`, `:long`
    and `:full`. The default is `:medium`.

  ## Example

      => Cldr.Strftime.strftime_options!
      [
        am_pm_names: #Function<0.117825700/1 in MyApp.Cldr.Calendar.strftime_options!/2>,
        month_names: #Function<1.117825700/1 in MyApp.Cldr.Calendar.strftime_options!/2>,
        abbreviated_month_names: #Function<2.117825700/1 in MyApp.Cldr.Calendar.strftime_options!/2>,
        day_of_week_names: #Function<3.117825700/1 in MyApp.Cldr.Calendar.strftime_options!/2>,
        abbreviated_day_of_week_names: #Function<4.117825700/1 in MyApp.Cldr.Calendar.strftime_options!/2>,
        preferred_date: "%b %d, %Y",
        preferred_time: "%H:%M:%S %p",
        preferred_datetime: "%b %d, %Y, %H:%M:%S %p"
      ]

  ## Typical usage from Elixir 1.11

      iex> Calendar.strftime ~U[2019-08-26 13:52:06.0Z], "%x", Cldr.Strftime.strftime_options!
      "Aug 26, 2019"

      iex> Calendar.strftime ~U[2019-08-26 13:52:06.0Z], "%X", Cldr.Strftime.strftime_options!
      "13:52:06 PM"

      iex> Calendar.strftime ~U[2019-08-26 13:52:06.0Z], "%c", Cldr.Strftime.strftime_options!
      "Aug 26, 2019, 13:52:06 PM"

  """
  def strftime_options!(locale \\ Cldr.get_locale(), options \\ [])

  def strftime_options!(options, []) when is_list(options) do
    {locale, _backend} = Cldr.locale_and_backend_from(options)
    strftime_options!(locale, options)
  end

  def strftime_options!(locale, options) when is_list(options) do
    {locale, backend} = Cldr.locale_and_backend_from(locale, Keyword.get(options, :backend))

    with {:ok, locale} <- Cldr.validate_locale(locale, backend) do
      calendar_backend = Module.concat(locale.backend, Calendar)
      strftime_backend = Module.concat(locale.backend, Strftime)
      format = Keyword.get(options, :format, :medium)

      locale
      |> calendar_backend.strftime_options!(options)
      |> Keyword.merge(preferred_formats(locale, format, strftime_backend))
    else
      {:error, {exception, message}} -> raise exception, message
    end
  end

  def preferred_formats(locale, format, strftime_backend) do
    [
      preferred_date:
        locale
        |> strftime_backend.date_formats()
        |> unwrap_ok!
        |> Map.fetch!(format),
      preferred_time:
        locale
        |> strftime_backend.time_formats()
        |> unwrap_ok!
        |> Map.fetch!(format),
      preferred_datetime:
        locale
        |> strftime_backend.date_time_formats()
        |> unwrap_ok!
        |> Map.fetch!(format)
    ]
  end

  defp unwrap_ok!({:ok, term}), do: term
  defp unwrap_ok!({:error, {exception, message}}), do: raise(exception, message)
end
